spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/SparkHilbertUDF.java

```diff
@@ -0,0 +1,243 @@
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.types.BinaryType;
+import org.apache.spark.sql.types.BooleanType;
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.DateType;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.DoubleType;
+import org.apache.spark.sql.types.FloatType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.TimestampType;
+import org.davidmoten.hilbert.HilbertCurve;
+import scala.collection.JavaConverters;
+import scala.collection.Seq;
+
+class SparkHilbertUDF implements SparkSpaceCurveUDF, Serializable {
+  private static final long PRIMITIVE_EMPTY = Long.MAX_VALUE;
+
+  private static final int BITS_NUM = 63;
+
+  SparkHilbertUDF() {}
+
+  private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+    in.defaultReadObject();
+  }
+
+  byte[] hilbertCurvePosBytes(Seq<Long> points) {
+    java.util.List<Long> longs = JavaConverters.seqAsJavaList(points);
+    long[] longs1 = new long[longs.size()];
+    for (int i = 0; i < longs.size(); i++) {
+      longs1[i] = longs.get(i);
+    }
+    HilbertCurve hilbertCurve = HilbertCurve.bits(BITS_NUM).dimensions(points.size());
+    return HilbertCurveUtils.indexBytes(hilbertCurve, longs1, BITS_NUM);
+  }
+
+  private UserDefinedFunction tinyToOrderedLongUDF() {
+    UserDefinedFunction udf =
+        functions
+            .udf(
+                (Byte value) -> {
+                  if (value == null) {
+                    return PRIMITIVE_EMPTY;
+                  }
+                  return BinaryUtil.convertBytesToLong(new byte[] {value});
+                },
+                DataTypes.LongType)
+            .withName("TINY_ORDERED_BYTES");
+
+    return udf;
+  }
+
+  private UserDefinedFunction shortToOrderedLongUDF() {
+    UserDefinedFunction udf =
+        functions
+            .udf(
+                (Short value) -> {
+                  if (value == null) {
+                    return PRIMITIVE_EMPTY;
+                  }
+                  return (long) value;
+                },
+                DataTypes.LongType)
+            .withName("SHORT_ORDERED_BYTES");
+
+    return udf;
+  }
+
+  private UserDefinedFunction intToOrderedLongUDF() {
+    UserDefinedFunction udf =
+        functions
+            .udf(
+                (Integer value) -> {
+                  if (value == null) {
+                    return PRIMITIVE_EMPTY;
+                  }
+                  return (long) value;
+                },
+                DataTypes.LongType)
+            .withName("INT_ORDERED_BYTES");
+
+    return udf;
+  }
+
+  private UserDefinedFunction longToOrderedLongUDF() {
```

Comment:
These are all the same as the Zorder functions correct? Can we unify the byte transformation udfs?

Reply:
yes,we can, I thought if I converted to bytes, I'd have to go back to long.